bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java

```diff
@@ -724,12 +807,42 @@ public ByteBuf internalReadEntry(long ledgerId, long entryId, long location)
         }
         if (entrySize < MIN_SANE_ENTRY_SIZE) {
             LOG.error("Read invalid entry length {}", entrySize);
-            throw new IOException("Invalid entry length " + entrySize);
+            throw new EntryLookupException.InvalidEntryLengthException(ledgerId, entryId, entryLogId, pos);
+        }
+
+        long thisLedgerId = sizeBuff.getLong(4);
+        long thisEntryId = sizeBuff.getLong(12);
+        if (thisLedgerId != ledgerId || thisEntryId != entryId) {
+            throw new EntryLookupException.WrongEntryException(
```

Comment:
This is problematic because it happens even from the `internalReadEntry()`. The purpose of `internalReadEntry()` was to avoid the sanity checks from the regular `readEntry()` and to be called when doing read-ahead. 

In read-ahead mode, we don't know whether the next entry is from the same ledger or not. Therefore we keep reading until we find an entry of a different ledger. 

With this change in 4.9.2 that means each read-ahead incurs in creating an exception, which is very expensive (since it will walk through the stack trace each time).